warp10/src/main/java/io/warp10/continuum/gts/INTERPOLATE.java

```diff
@@ -27,23 +25,514 @@
+      function = (new LinearInterpolator()).interpolate(xval, fval);
+    } else {
+      switch (Interpolator.valueOf((String) params.get(PARAM_INTERPOLATOR))) {
+        case spline:
+          if (nvalues > 2) {
+            function = (new SplineInterpolator().interpolate(xval, fval));
+            break;
+          }
+        case akima:
+          if (nvalues > 4 ) {
+            function = (new AkimaSplineInterpolator().interpolate(xval, fval));
+            break;
+          }
+        case linear:
+          if (nvalues > 1) {
+            function = (new LinearInterpolator()).interpolate(xval, fval);
+            break;
+          }
+        case noop:
+          function = null;
+          break;
+      }
+    }
+
+    //
+    // Fill the result
+    // Count ticks with elevation and location
+    //
+
+    int nElevations = 0;
+    int nLocations = 0;
+
+    if (null == params.get(PARAM_OCCURRENCES)) {
+      // in this case, the result is necessarily bucketized
+      // we fill empty buckets
+
+      //
+      // Compute oldest bucket
+      //
+
+      long bucket = filled.lastbucket - filled.bucketcount * filled.bucketspan;
+
+      if ((null != filled.elevations) && (GeoTimeSerie.NO_ELEVATION != filled.elevations[0])) {
+        nElevations++;
+      }
+
+      if ((null != filled.locations) && (GeoTimeSerie.NO_LOCATION != filled.locations[0])) {
+        nLocations++;
+      }
+
+      for (int i = 1; i < nvalues; i++) {
+        if ((null != filled.elevations) && (GeoTimeSerie.NO_ELEVATION != filled.elevations[i])) {
+          nElevations++;
+        }
+        if ((null != filled.locations) && (GeoTimeSerie.NO_LOCATION != filled.locations[i])) {
+          nLocations++;
+        }
+
+        //
+        // Move bucket passed the last tick encountered
+        //
+        while(bucket < filled.lastbucket && bucket <= filled.ticks[i-1]) {
+          bucket += filled.bucketspan;
+        }
+
+        //
+        // If bucket is on the current tick, advance tick
+        //
+        if (bucket == filled.ticks[i]) {
+          continue;
+        }
+
+        //
+        // Fill missing values until bucket passes the current tick
+        //
+
+        while(bucket < filled.ticks[i]) {
+          if (function.isValidPoint(bucket)) {
+            GTSHelper.setValue(filled, bucket, function.value(bucket));
+          } else {
+            Number filler = (Number) params.get(PARAM_INVALID_TICK_VALUE);
+            if (null != filler) {
+              GTSHelper.setValue(filled, bucket, filler);
+            }
+          }
+          bucket += filled.bucketspan;
+        }
+      }
+
+    } else {
+      // count geo
+      for (int i = 0; i < nvalues; i++) {
+
+        if ((null != filled.elevations) && (GeoTimeSerie.NO_ELEVATION != filled.elevations[i])) {
+          nElevations++;
+        }
+
+        if ((null != filled.locations) && (GeoTimeSerie.NO_LOCATION != filled.locations[i])) {
+          nLocations++;
+        }
+      }
```

Comment:
If my understanding is correct, this can be done once before the occurences/bucket choice. From the result, you can build interpolation functions for location and elevation (or leave a null), then you can iterate just once on all occurrences/bucket.